paddle/phi/kernels/sparse/gpu/reshape_kernel.cu

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/phi/kernels/sparse/unary_kernel.h"
+
+#include "paddle/phi/backends/gpu/gpu_context.h"
+#include "paddle/phi/core/kernel_registry.h"
+#include "paddle/phi/kernels/funcs/elementwise_base.h"
+#include "paddle/phi/kernels/sparse/empty_kernel.h"
+#include "paddle/phi/kernels/sparse/impl/unary_kernel_impl.h"
+
+#include "paddle/phi/kernels/sparse/sparse_utils_kernel.h"
+
+
+namespace phi {
+namespace sparse {
+
+__global__ void ReshapeCooCudaKernel(const int64_t* x_indices_data,
+                                     const int num_x_sparse_part_dims,
+                                     const int num_out_sparse_part_dims,
+                                     const int64_t x_nnz,
+                                     const int64_t* x_sparse_part_strides,
+                                     const int64_t* out_sparse_part_strides,
+                                     int64_t *out_indices_data) {
+
+    CUDA_KERNEL_LOOP_TYPE(j, x_nnz, int64_t) {
+        int64_t location = 0;
+        for (int i = 0; i < num_x_sparse_part_dims; ++i) {
+            location += x_indices_data[i * x_nnz + j] * x_sparse_part_strides[i];
+            // row major or column major ???
+            // location += x_indices_data[j * num_x_sparse_part_dims + i] * x_sparse_part_strides[i];
+        }
+        for (int i = 0; i < num_out_sparse_part_dims; ++i) {
+            out_indices_data[i * x_nnz + j] = location / out_sparse_part_strides[i];
+            // row major or column major ???
+            // out_indices_data[j * num_out_sparse_part_dims + i] = location / out_sparse_part_strides[i];
+            location %= out_sparse_part_strides[i];
+        }
+    }
+}
+
+
+template <typename T, typename Context>
+void ReshapeCooKernel(const Context& dev_ctx,
+                      const SparseCooTensor &x,
+                      const phi::IntArray& shape,
+                      SparseCooTensor *out) {
+  int64_t x_nnz = x.nnz();
+  // TODO: consider using "DDim DDim::reshape(std::vector<int>& shape)"
+  // DDim out_dims = phi::make_ddim(shape.GetData());
+  DDim out_dims = x.dims().reshape(std::vector<int>(shape.GetData().begin(), shape.GetData().end()));
```

Comment:
This line seems to report an error:
<img width="885" alt="image" src="https://user-images.githubusercontent.com/31559413/194830935-17fa4b97-306e-4fb6-a26c-100229ba9d58.png">

Solution: change it to the following:
```c++
  std::vector<int> new_shape(shape.GetData().begin(), shape.GetData().end());
  phi::DDim out_dims = x.dims().reshape(new_shape);
```